app/src/main/java/net/squanchy/notification/NotificationWorker.kt

```diff
@@ -0,0 +1,120 @@
+
+        val notificationIntervalEnd = now.plusMinutes(NOTIFICATION_INTERVAL_MINUTES)
+
+        val sortedFavourites = service.sortedFavourites()
+
+        val showNotification = if (shouldShowNotifications()) {
+            sortedFavourites
+                .map { events -> events.filter { it.startTime.isAfter(now) } }
+                .map { events -> events.filter { isBeforeOrEqualTo(it.startTime, notificationIntervalEnd) } }
+                .map(notificationCreator::createFrom)
+                .doOnSuccess(notifier::showNotifications)
+                .doOnError(::logError)
+                .map { Result.success() }
+                .onErrorReturnItem(Result.failure())
+        } else {
+            Single.just(Result.success())
+        }
+
+        val scheduleNext = sortedFavourites
+            .map { events -> events.filter { it.startTime.isAfter(notificationIntervalEnd) } }
+            .doOnSuccess(::scheduleNextAlarm)
+            .doOnError(::logError)
+            .map { Result.success() }
+            .onErrorReturnItem(Result.failure())
+
+        return Single.zip(
+            showNotification,
+            scheduleNext,
+            BiFunction<Result, Result, Result> { notificationResult, scheduleResult ->
+                if (notificationResult is Result.Failure || scheduleResult is Result.Failure) {
+                    Result.failure()
+                } else {
+                    Result.success()
+                }
+            }
+
+        )
+    }
+
+    private fun logError(error: Throwable) {
+        Timber.e(error, "Error in NotificationsIntentService: ${error.javaClass}")
+    }
+
+    private fun shouldShowNotifications(): Boolean {
+        val notificationPreferenceKey = applicationContext.getString(R.string.about_to_start_notification_preference_key)
+        return preferences.getBoolean(notificationPreferenceKey, SHOW_NOTIFICATIONS_DEFAULT)
+    }
+
+    private fun isBeforeOrEqualTo(start: LocalDateTime, notificationIntervalEnd: LocalDateTime): Boolean {
+        return start.isBefore(notificationIntervalEnd) || start.isEqual(notificationIntervalEnd)
+    }
+
+    private fun scheduleNextAlarm(events: List<Event>) {
+        if (events.isEmpty()) {
+            Timber.d("No future events to schedule an alarm for.")
+            return
+        }
+        val startTime = events[0].startTime
+        val serviceAlarm = startTime.minusMinutes(NOTIFICATION_INTERVAL_MINUTES)
+
+        Timber.d("Next alarm scheduled for %s", serviceAlarm.toString())
+
+        scheduleNotificationWork(
+            Duration.between(
+                currentTime.currentDateTime(),
+                serviceAlarm
+            )
+        )
+    }
+
+    companion object {
+        private const val NOTIFICATION_INTERVAL_MINUTES = 10L
+        private const val SHOW_NOTIFICATIONS_DEFAULT = true
+    }
+}
+
+fun scheduleNotificationWork(delay: Duration = Duration.ZERO) {
```

Comment:
I decided to go for a duration over a date time here.

The reason behind that is that the api we're using is indeed working with a delay, which is better represented as a duration. The initial idea I had was to pass a date time as a parameter, but that requires us to

1. Get a `CurrentTime` instance here somehow (to honor our clock)
2. Calculate the time difference between now and the target time

`1` is problematic because then you'd have to expose that dependency as a parameter, which is fine but I found it annoying.

`2` can result in a weird behavior: if you pass `now` here as the target time, the difference is at best 0, at worst negative (because you need to get the current `now` which is at best the same as the one passed).

Given all the usage are (apart from inside this file) to start the work immediately, the default value of zero is actually used and makes everything simpler imho.